src/renderer/base/RenderClusterIterator.cpp

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "precomp.h"
+#include "../inc/RenderClusterIterator.hpp"
+
+#include "../../buffer/out/textBufferCellIterator.hpp"
+
+using namespace Microsoft::Console::Render;
+
+RenderClusterIterator::RenderClusterIterator(TextBufferCellIterator& cellIter) :
+    _cellIter(cellIter),
+    _cluster(L"", 0),
+    _attr((*cellIter).TextAttr()),
+    _distance(0),
+    _exceeded(false)
+{
+    _GenerateCluster();
+}
+
+RenderClusterIterator::operator bool() const noexcept
+{
+    return !_exceeded;
+}
+
+// Routine Description:
+// - Compares two iterators to see if they're pointing to the same position in the same buffer
+// Arguments:
+// - it - The other iterator to compare to this one.
+// Return Value:
+// - True if it's the same text buffer and same cell position. False otherwise.
+bool RenderClusterIterator::operator==(const RenderClusterIterator& it) const
+{
+    return _attr == it._attr &&
+           &_cellIter == &it._cellIter &&
+           &_distance == &it._distance &&
+           _exceeded == it._exceeded;
+}
+
+// Routine Description:
+// - Compares two iterators to see if they're pointing to the different positions in the same buffer or different buffers entirely.
+// Arguments:
+// - it - The other iterator to compare to this one.
+// Return Value:
+// - True if it's the same text buffer and different cell position or if they're different buffers. False otherwise.
+bool RenderClusterIterator::operator!=(const RenderClusterIterator& it) const
+{
+    return !(*this == it);
+}
+
+RenderClusterIterator& RenderClusterIterator::operator+=(const ptrdiff_t& movement)

```

Comment:
Currently `RenderClusterIterator` is not aware of the `dbcs` columns. Caller must use `clusterIter += cols > 0 ? cols : 1` to skip dbcs trailing char. I've tried to let `RenderClusterIterator` handle this itself, but failed to do so because the code gets too complicated and implicit.